8354887: Preserve runtime blobs in AOT code cache

ashu-mehra opened the pull request:
[8350209](https://bugs.openjdk.org/browse/JDK-8350209) introduced the framework for storing code in aot code cache and used it for caching i2c/c2i adapters.
This PR extends the `AOTCodeCache` infrastructure and stores various runtime blobs (shared blobs, C1 and C2 runtime blobs) in the AOT code cache. It adds a new diagnostic flag `AOTStubCaching` to enable/disable the caching of these blobs.
`AOTCodeFlags.java` test is extended to cover `AOTStubCaching`.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8354887](https://bugs.openjdk.org/browse/JDK-8354887): Preserve runtime blobs in AOT code cache (**Enhancement** - P4)


### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)
 * [Andrew Dinn](https://openjdk.org/census#adinn) (@adinn - **Reviewer**)

### Contributors
 * Andrew Dinn `<adinn@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/25019/head:pull/25019` \
`$ git checkout pull/25019`

Update a local copy of the PR: \
`$ git checkout pull/25019` \
`$ git pull https://git.openjdk.org/jdk.git pull/25019/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 25019`

View PR using the GUI difftool: \
`$ git pr show -t 25019`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/25019.diff">https://git.openjdk.org/jdk/pull/25019.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/25019#issuecomment-2848420639)
</details>

ashu-mehra commented:
/cc hotspot-runtime
/cc hotspot-compiler

/label remove hotspot

vnkozlov commented on src/hotspot/share/code/aotCodeCache.cpp:
```diff
@@ -22,6 +22,9 @@
  *
  */
 
+#ifdef COMPILER1
+#include "c1/c1_Runtime1.hpp"
+#endif
```
Conditional includes are placed at the end of includes.

vnkozlov commented on src/hotspot/share/code/aotCodeCache.cpp:
```diff
@@ -348,10 +408,18 @@ bool AOTCodeCache::Config::verify() const {
     log_debug(aot, codecache, init)("AOT Code Cache disabled: it was created with CompressedOops::shift() = %d vs current %d", _compressedOopShift, CompressedOops::shift());
     return false;
   }
+  if (_compressedOopBase != CompressedOops::base()) {
+    log_debug(aot, codecache, init)("AOT Code Cache disabled: it was created with CompressedOops::base() = %p vs current %p", _compressedOopBase, CompressedOops::base());
+    return false;
+  }
```
I think we have relocation for CompressedOops::base() so we can patch. No need to bailout.
Do you have stub/blob which missed relocation?

ashu-mehra replied:
> I think we have relocation for CompressedOops::base() so we can patch.

We have relocation for `CompressedOops::base()` only when heap is not yet initialized:

```
void MacroAssembler::reinit_heapbase() {
  if (UseCompressedOops) {
    if (Universe::heap() != nullptr) {
      if (CompressedOops::base() == nullptr) {
        MacroAssembler::xorptr(r12_heapbase, r12_heapbase);
      } else {
        mov64(r12_heapbase, (int64_t)CompressedOops::base());
      }
    } else {
      movptr(r12_heapbase, ExternalAddress(CompressedOops::base_addr()));
    } 
  } 
}
```

In premain we do add `CompressedOops::base_addr()` to the AOT Address table. But I don't think we are accessing `CompressedOops::base` in adapters or blobs that we are targeting in mainline. At least I haven't across the need to have `CompressedOops::base` in the address table. Is that correct @vnkozlov @adinn ?

vnkozlov commented on src/hotspot/share/code/aotCodeCache.cpp:
```diff
@@ -348,10 +408,18 @@ bool AOTCodeCache::Config::verify() const {
     log_debug(aot, codecache, init)("AOT Code Cache disabled: it was created with CompressedKlassPointers::shift() = %d vs current %d", _compressedKlassShift, CompressedKlassPointers::shift());
     return false;
   }
+  if (_compressedKlassBase != CompressedKlassPointers::base()) {
+    log_debug(aot, codecache, init)("AOT Code Cache disabled: it was created with CompressedKlassPointers::base() = %p vs current %p", _compressedKlassBase, CompressedKlassPointers::base());
+    return false;
+  }
```
I would suggest to use relocation for klass's base too but not in these changes. I think bailout is fine here.

vnkozlov commented on src/hotspot/share/code/aotCodeCache.cpp:
```diff
@@ -663,10 +765,13 @@ bool AOTCodeCache::store_code_blob(CodeBlob& blob, AOTCodeEntry::Kind entry_kind
     return false;
   }
-  log_debug(aot, codecache, stubs)("Writing blob '%s' to AOT Code Cache", name);
+  log_info(aot, codecache, stubs)("Writing blob '%s' (id=%u, kind=%s) to AOT Code Cache", name, id, AOTCodeEntry::kind_string(entry_kind));
```
Please, use `log_debug()` in final changes.

vnkozlov commented on src/hotspot/share/code/aotCodeCache.cpp:
```diff
@@ -747,17 +862,23 @@ CodeBlob* AOTCodeCache::load_code_blob(AOTCodeEntry::Kind entry_kind, uint id, c
-  return reader.compile_code_blob(name, entry_offset_count, entry_offsets);
+  CodeBlob* blob = reader.compile_code_blob(name, entry_offset_count, entry_offsets);
+
+  log_info(aot, codecache, stubs)("Read blob '%s' (id=%u, kind=%s) from AOT Code Cache", name, id, AOTCodeEntry::kind_string(entry_kind));
```
Use `log_debug()`

vnkozlov commented on src/hotspot/share/code/aotCodeCache.cpp:
```diff
@@ -965,6 +1104,82 @@ ImmutableOopMapSet* AOTCodeReader::read_oop_map_set() {
+  }
+  uint count = 0;
+  bool result = cb.asm_remarks().iterate([&] (uint offset, const char* str) -> bool {
+    log_trace(aot, codecache, stubs)("asm remark offset=%d, str='%s'", offset, str);
+    uint n = write_bytes(&offset, sizeof(uint));
+    if (n != sizeof(uint)) {
+      return false;
```
Consider using `id_for_C_string()`  and record ID instead of coping string. These strings should be recorded in C strings table already.
If `id_for_C_string()` does not find - assert. We should add `add_C_string()` in missing place.

ashu-mehra replied:
The asm remarks and dbg strings are not currently recorded in C string table.
I tried to add these strings by calling `AOTCodeCache::add_C_string()`  in `AsmRemarkCollection::insert()` but this results in adding LOTS of strings. So I add the strings to the string table only when writing the asm remarks. This keeps the string count in check.

vnkozlov commented on src/hotspot/share/code/aotCodeCache.cpp:
```diff
@@ -965,6 +1104,82 @@ ImmutableOopMapSet* AOTCodeReader::read_oop_map_set() {
+  }
+  uint count = 0;
+  bool result = cb.dbg_strings().iterate([&] (const char* str) -> bool {
+    log_trace(aot, codecache, stubs)("dbg string=%s", str);
+    uint len = (uint)strlen(str) + 1; // including '\0' char
+    uint n = write_bytes(str, len);
```
Same here

vnkozlov reviewed (commented):
Few comments.

vnkozlov commented:
We need to do something about Compressed Klass base:
```
java -XX:AOTCache=app.aotcache -XX:+UnlockDiagnosticVMOptions -XX:+AOTAdapterCaching -XX:+AOTStubCaching -Xlog:aot+codecache+init=debug -cp hello.jar HelloWorld
[0.009s][debug][aot,codecache,init] Mapped 548864 bytes at address 0x00007f17b1611000 at AOT Code Cache
[0.009s][info ][aot,codecache,init] Loaded 384 AOT code entries from AOT Code Cache
[0.009s][debug][aot,codecache,init]   Adapters:  total=316
[0.009s][debug][aot,codecache,init]   Shared Blobs: total=14
[0.009s][debug][aot,codecache,init]   C1 Blobs: total=34
[0.009s][debug][aot,codecache,init]   C2 Blobs: total=20
[0.009s][debug][aot,codecache,init]   AOT code cache size: 543392 bytes
[0.009s][debug][aot,codecache,init]   Loaded 1 C strings of total length 28 at offset 521860 from AOT Code Cache
[0.010s][debug][aot,codecache,init] AOT Code Cache disabled: it was created with CompressedKlassPointers::base() = 0x7f56c6000000 vs current 0x7f1762000000
[0.010s][info ][aot,codecache,init] Unable to use AOT Code Cache.
Hellow World!
```

vnkozlov commented:
Which blob/stubs decompress/compress klass using the base?

vnkozlov commented:
May be we should use Relocation for it.

ashu-mehra commented:
@vnkozlov How about updating `decode_klass_not_null` like this?

```
  if (CompressedKlassPointers::base() != nullptr) {
    if (AOTCodeCache::is_on_for_dump()) {
      movptr(tmp, ExternalAddress(CompressedKlassPointers::base_addr()));
    } else {
      mov64(tmp, (int64_t)CompressedKlassPointers::base());
    }
    addq(r, tmp);
  }
```

and adding `CompressedKlassPointers::base_addr()` to the AOTAddressTable.

vnkozlov commented:
Yes, please do that in `decode_klass_not_null()`.
We update other methods for nmethod caching in JDK 26.

vnkozlov commented:
I think `reinit_heapbase()` is used after JNI calls to put it back into register for compiled code. I don't think 
For debugging your changes you can add print Into `reinit_heapbase()` using UseNewCode flag and set the flag in  `compiler_stubs_init()`

ashu-mehra commented:
/contributor add @adinn

vnkozlov commented:
This looks good. Let me test it.

vnkozlov commented:
1% improvement ;^)
```
ghost29:jdk_git2$ (perf stat -r 100 ./build/product/images/jdk/bin/java -XX:AOTCache=app.aotcache -XX:+UnlockDiagnosticVMOptions -XX:-AOTAdapterCaching -XX:-AOTStubCaching -cp hello.jar HelloWorld > /dev/null) 2>&1 | grep elapsed
         0.0252305 +- 0.0000543 seconds time elapsed  ( +-  0.22% )
ghost29:jdk_git2$ (perf stat -r 100 ./build/product/images/jdk/bin/java -XX:AOTCache=app.aotcache -XX:+UnlockDiagnosticVMOptions -XX:+AOTAdapterCaching -XX:-AOTStubCaching -cp hello.jar HelloWorld > /dev/null) 2>&1 | grep elapsed
         0.0234828 +- 0.0000415 seconds time elapsed  ( +-  0.18% )
ghost29:jdk_git2$ (perf stat -r 100 ./build/product/images/jdk/bin/java -XX:AOTCache=app.aotcache -XX:+UnlockDiagnosticVMOptions -XX:+AOTAdapterCaching -XX:+AOTStubCaching -cp hello.jar HelloWorld > /dev/null) 2>&1 | grep elapsed
         0.0232267 +- 0.0000355 seconds time elapsed  ( +-  0.15% )
```

vnkozlov commented:
Test crashes on linux-x64 with debug VM:
```
% make test JTREG=AOT_JDK=true CONF=fastdebug TEST=compiler/c2/cr6865031/Test.java
```

vnkozlov commented:
Test `runtime/cds/appcds/aotClassLinking/MethodHandleTest.java` crashed on linux-aarch64:

```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x0000ffffb4df1f84, pid=3645763, tid=3645766
#
# JRE version: Java(TM) SE Runtime Environment (25.0) (fastdebug build 25-internal-LTS-2025-05-05-2218263.vladimir.kozlov.jdkgit2)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (fastdebug 25-internal-LTS-2025-05-05-2218263.vladimir.kozlov.jdkgit2, mixed mode, sharing, tiered, compressed oops, compressed class ptrs, g1 gc, linux-aarch64)
# Problematic frame:
# v  ~RuntimeStub::Shared Runtime throw_NullPointerException_at_call_blob 0x0000ffff73dd2db0
```

vnkozlov commented:
I attached hs_err file to RFE in JBS.

ashu-mehra commented:
@vnkozlov thanks for testing the patch.

> % make test JTREG=AOT_JDK=true CONF=fastdebug TEST=compiler/c2/cr6865031/Test.java

I can recreate this test locally. Looking into it.

ashu-mehra commented:
> I can recreate this test locally. Looking into it.

Looks like the code for C2 blobs generated in assembly phase is not correct.
For example, code for new_instance blob is:

```
[0.141s][550409][aot,codecache,stubs] Decoding CodeBlob, name: C2 Runtime new_instance, at  [0x00007fed0f904260, 0x00007fed0f9042b8]  88 bytes
[0.141s][550409][aot,codecache,stubs]  ;; N1: # out( B1 ) <- in( B3 B2 )  Freq: 1
[0.141s][550409][aot,codecache,stubs]  ;; B1: # out( B3 B2 ) <- BLOCK HEAD IS JUNK  Freq: 1
[0.141s][550409][aot,codecache,stubs]   0x00007fed0f904260:   sub    $0x8,%rsp
[0.141s][550409][aot,codecache,stubs]   0x00007fed0f904267:   mov    %rbp,(%rsp)
[0.141s][550409][aot,codecache,stubs]   0x00007fed0f90426b:   mov    %rsp,0x3e8(%r15)
[0.141s][550409][aot,codecache,stubs]   0x00007fed0f904272:   mov    %rsi,%rdi
[0.141s][550409][aot,codecache,stubs]   0x00007fed0f904275:   mov    %r15,%rsi
[0.141s][550409][aot,codecache,stubs]   0x00007fed0f904278:   movabs $0x7fed28908982,%r10
[0.141s][550409][aot,codecache,stubs]   0x00007fed0f904282:   call   *%r10
[0.141s][550409][aot,codecache,stubs]   0x00007fed0f904285:   nopl   0x0(%rax,%rax,1)
[0.141s][550409][aot,codecache,stubs]   0x00007fed0f90428d:   mov    %r12,0x3e8(%r15)
[0.141s][550409][aot,codecache,stubs]   0x00007fed0f904294:   mov    %r12,0x3f0(%r15)
[0.141s][550409][aot,codecache,stubs]   0x00007fed0f90429b:   mov    0x440(%r15),%rax
[0.141s][550409][aot,codecache,stubs]   0x00007fed0f9042a2:   mov    %r12,0x440(%r15)
[0.141s][550409][aot,codecache,stubs]   0x00007fed0f9042a9:   cmp    0x8(%r15),%r12
[0.141s][550409][aot,codecache,stubs]   0x00007fed0f9042ad:   jne    0x00007fed0f9042b1
[0.141s][550409][aot,codecache,stubs]  ;; B2: # out( N1 ) <- in( B1 )  Freq: 0.999999
[0.141s][550409][aot,codecache,stubs]   0x00007fed0f9042af:   pop    %rbp
[0.141s][550409][aot,codecache,stubs]   0x00007fed0f9042b0:   ret    
[0.141s][550409][aot,codecache,stubs]  ;; B3: # out( N1 ) <- in( B1 )  Freq: 1e-06
[0.141s][550409][aot,codecache,stubs]   0x00007fed0f9042b1:   pop    %rbp
[0.141s][550409][aot,codecache,stubs]   0x00007fed0f9042b2:   jmp    Stub::forward_exception
[0.141s][550409][aot,codecache,stubs]   0x00007fed0f9042b7:   hlt    
```
Look at the instructions generated after the call to runtime method:

```
[0.141s][550409][aot,codecache,stubs]   0x00007fed0f90428d:   mov    %r12,0x3e8(%r15)
[0.141s][550409][aot,codecache,stubs]   0x00007fed0f904294:   mov    %r12,0x3f0(%r15)
[0.141s][550409][aot,codecache,stubs]   0x00007fed0f90429b:   mov    0x440(%r15),%rax
[0.141s][550409][aot,codecache,stubs]   0x00007fed0f9042a2:   mov    %r12,0x440(%r15)
[0.141s][550409][aot,codecache,stubs]   0x00007fed0f9042a9:   cmp    0x8(%r15),%r12
[0.141s][550409][aot,codecache,stubs]   0x00007fed0f9042ad:   jne    0x00007fed0f9042b1
```

Comparing it with the case when AOT code cache is not used:
```
   0x7fffdef07c8d:	movq   $0x0,0x3e8(%r15)
   0x7fffdef07c98:	movq   $0x0,0x3f0(%r15)
   0x7fffdef07ca3:	mov    0x440(%r15),%rax
   0x7fffdef07caa:	movq   $0x0,0x440(%r15)
   0x7fffdef07cb5:	mov    0x8(%r15),%r10
   0x7fffdef07cb9:	test   %r10,%r10
   0x7fffdef07cbc:	jne    0x7fffdef07cc0
```

For some reason r12 is used instead of null (0x0). r12 is the CompressedOop::base address. 
C2 code that generates this is https://github.com/openjdk/jdk/blob/762423d64d10dcdb37800767d2b2f1b7757c804a/src/hotspot/share/opto/generateOptoStub.cpp#L222

I suspect I missed porting a change from premain. @adinn @vnkozlov any idea what that could be?

vnkozlov commented:
Easy to fix. Look for checks `UseCompressedOops && (CompressedOops::base() == nullptr)` in predicates in `x86_64.ad` and add additional check `!AOTCodeCache::is_on_for dump()`. May be create new function `r12_is_null()` or something.

adinn commented:
> I suspect I missed porting a change from premain. @adinn @vnkozlov any idea what that could be?

@ashu-mehra Just to explain what is going on here:

This is a performance trick. When compressed oops base is null r12 (aka rheapbase) will have been initialized to zero so it can be used as a zero register. This allows, for example, a move instruction to employ a register operand immediate rather than include a 64 bit zero value in the instruction stream, which results in reduced code size. In this case the two moves are zeroing the current Java thread's frame anchor fields, last Java frame pc and sp, which are only set while the thread is in native.

This trick is fine when generated code is run in the same JVM but no use if the code is generated in a VM with zero compressed oops base then reloaded into a JVM where it is no longer null.

So, Vladimir's advice is to disable this trick when generating AOT code.

vnkozlov commented:
In `premain` branch we don't have null in R12 because we set shift together with base when we generate AOT code: [compressedOops.cpp#L51](https://github.com/openjdk/leyden/blob/premain/src/hotspot/share/oops/compressedOops.cpp#L51)

ashu-mehra commented:
> It is still not enabled - how we are not crashing in premain?

That's a news to me. I thought we were setting `UseCompatibleCompressedOops` to true which is why we are avoiding this issue in premain. But ..

adinn commented:
@vnkozlov Ergonomics comes into play. It is set to true in cdsConfig.cpp if CacheDataStore != nullptr.

ashu-mehra commented:
@adinn in premain it is commented out -
https://github.com/openjdk/leyden/blob/f09d2f7724c628c90df51eacb16b33fee710ed1a/src/hotspot/share/cds/cdsConfig.cpp#L790

```
#ifdef _LP64
  // FLAG_SET_ERGO_IF_DEFAULT(UseCompatibleCompressedOops, true); // FIXME @iklam - merge with mainline - UseCompatibleCompressedOops
#endif
```

I don't see any other place it is set.

vnkozlov commented:
@adinn I think you are looking on old version of `premain` branch. The latest changeset is Igor's "Address review comments".

adinn commented:
Ah yes, I was looking at an old version. Apologies for that.

adinn commented:
I looked back at the history and found that this line got commented out as part of a merge from mainline (date: 2025-01-15 shaid 4cd4c7cdfae7b4b5eb3308abc8fa8d0ed7581ad8).

I'm not at all sure why Ioi did this and have no idea why he added my name to the original FIXME. The very next change he made in the premain branch was to update the FIXME comment, resetting it to his name.

I also don't understand why this is not biting us in premain.

vnkozlov commented:
There could be testing failures which expect some state of compressed oops. I will test enabling this flag.

vnkozlov commented:
We were lucky. I reproduced issue in `premain` with HelloWord by running `AOTMode=create -Xmx4G` and product run with `Xmx31g`. It has the same `shift = 3` but different base (0 for 4Gb).

vnkozlov commented:
Unfortunately we can't enable the flag (that is why Ioi commented it, I think):

#  Internal Error (/leyden/open/src/hotspot/share/oops/compressedOops.cpp:87), pid=1169225, tid=1169229
#  assert((intptr_t)base() <= ((intptr_t)_heap_address_range.start() - (intptr_t)os::vm_page_size()) || base() == nullptr) failed: invalid value
#
# JRE version:  (25.0) (fastdebug build )
# Java VM: Java HotSpot(TM) 64-Bit Server VM (fastdebug 25-internal-LTS-2025-05-07-1649004.vkozlov..., mixed mode, sharing, tiered, compressed oops, compressed class ptrs, g1 gc, linux-amd64)
# Problematic frame:
# V  [libjvm.so+0xba81e9]  CompressedOops::initialize(ReservedHeapSpace const&)+0x3c9

@ashu-mehra please restore verification check for COOP base you had. instead of patching `x86_64.ad` file

vnkozlov commented:
I think it is fine not use AOT when the heap base does not match. Based on my test it will happen only with big heap's size difference between assembly and production runs.

ashu-mehra commented:
@vnkozlov I wonder if that assert needs modification as well.

This assert assumes that if `CompressedOops::_base` is not null, then it will be set to a page size before the heap range start. This is fine because _base is not null only when heap range goes beyond `OopEncodingHeapMax`, and in such cases `ReservedHeapSpace::noaccess_prefix` is equal to `os::vm_page_size()`.
But with `UseCompatibleCompressedOops` the _base is set to non-null value even when heap range is within `OopEncodingHeapMax` and in such cases `ReservedHeapSpace::noaccess_prefix` is 0.

I think the assert should be using `noaccess_prefix` instead of hard-coding `os::vm_page_size`:
```
-  assert((intptr_t)base() <= ((intptr_t)_heap_address_range.start() - (intptr_t)os::vm_page_size()) ||
+  assert((intptr_t)base() <= ((intptr_t)_heap_address_range.start() - (intptr_t)heap_space.noaccess_prefix()) ||
```
With this change I can run the helloworld program with `UseCompatibleCompressedOops` enabled.

vnkozlov commented:
I also hit failure in `premain` testing in runtime/cds/appcds/leyden/LeydenHello.java#aot test on Windows-x64 due to wrong Compress Class encoding due to different base:
```
[0.031s][info][cds] The current max heap size = 1024M, G1HeapRegion::GrainBytes = 1048576
[0.031s][info][cds]     narrow_klass_base = 0x0000015298000000, arrow_klass_pointer_bits = 32, narrow_klass_shift = 0

AOT code uses 0x800000000 instead :

88: 8b 78 08                mov    edi,DWORD PTR [rax+0x8]
8b: 49 ba 00 00 00 00 08    movabs r10,0x800000000
92: 00 00 00
95: 49 03 fa                add    rdi,r10
98: 48 3b 5f 38             cmp    rbx,QWORD PTR [rdi+0x38]
```
So we need the relocation fix from current changes in `premain` branch too.

vnkozlov commented:
> With this change I can run the helloworld program with UseCompatibleCompressedOops enabled.

I will try your assert fix.

vnkozlov commented:
Several AOT tests failed on linux-x64 with flag enabled:
```
$ make test CONF=fast TEST=test/hotspot/jtreg/runtime/cds/appcds/aotClassLinking TEST_OPTS_JAVA_OPTIONS="-Xmixed -XX:+UseCompatibleCompressedOops"
...
   TEST                                              TOTAL  PASS  FAIL ERROR  SKIP   
   jtreg:open/test/hotspot/jtreg/runtime/cds/appcds/aotClassLinking
>>                                                      15     5     8     0     2
```

vnkozlov commented:
I moved `UseCompatibleCompressedOops` flag setting to `check_vm_args_consistency()` to affect current workflow.
And I used `if (AOTCodeCache::is_caching_enabled()) {` for its setting.

vnkozlov commented:
I think for these changes we should not use AOT code when the heap base does not match.
Something changed in compressed oops code which prevents enforcing encoding.
We can investigate and fix it later.

vnkozlov commented:
@rose00 We have solution for that: we enforce encoding instructions (UseCompatibleCompressedOops flag) and use relocation info to patch correct COOP base address when loading AOT code.
The current issue is something changed in COOP code in runtime which cause next assert hit when we enforce encoding and use AOT code: [instanceKlass.cpp#L794](https://github.com/openjdk/leyden/blob/premain/src/hotspot/share/oops/instanceKlass.cpp#L794)

ashu-mehra commented:
> I think for these changes we should not use AOT code when the heap base does not match.
Something changed in compressed oops code which prevents enforcing encoding.
We can investigate and fix it later.

@vnkozlov for this PR we are relying on having relocation for COOP base, not on enforcing encoding. And that should be able to handle cases where heap base is different in assembly vs prod. Why do you suggest to not use AOT code when the heap base does not match?

vnkozlov commented:
To support that you would need to modify x86_64.ad to exclude r12 from usage as 0 as we discussed.
I think that using your original changes to not use AOT code with mismatched coop base is simpler for these changes and more robust for mainline. I concern that we may missing some places which will cause issues later. And we near JDK 25 fork. We can do more changes/improvements later in JDK 26.

One suggestion I have is to skip using only new runtime blobs you are adding when bases do not match because adapters don't have this issue.

adinn commented:
> One suggestion I have is to skip using only new runtime blobs you are adding when bases do not match because adapters don't have this issue.

c2i adapters do suffer this issue. They call out to the barrier set to plant  a c2i entry barrier. The barrier performs a weak reference load through field `ClassLoaderData::holder` (type `WeakHandle`). Method `resolve_weak_handle` calls `decode_heap_oop`.

vnkozlov commented on src/hotspot/share/code/aotCodeCache.hpp:
```diff
@@ -165,6 +166,7 @@ class AOTCodeCache : public CHeapObj<mtCode> {
 protected:
   class Config {
     uint _compressedOopShift;
+    address _compressedOopBase;
```
Put it fist to avoid gaps between fields.

vnkozlov commented on test/hotspot/jtreg/runtime/cds/appcds/aotCode/AOTCodeCompressedOopsTest.java:
```diff
@@ -0,0 +1,163 @@
+import jdk.test.lib.cds.CDSAppTester;
+import jdk.test.lib.process.OutputAnalyzer;
+
+public class AOTCodeCompressedOopsTest {
```
Ioi also suggested test example which I attached to RFE in JBS.
Looks like your test is very similar and more.

vnkozlov commented on src/hotspot/share/code/aotCodeCache.hpp:
```diff
@@ -331,6 +370,9 @@ class AOTCodeCache : public CHeapObj<mtCode> {
   static bool is_dumping_adapters() NOT_CDS_RETURN_(false);
   static bool is_using_adapters() NOT_CDS_RETURN_(false);
 
+  static bool is_dumping_stubs() NOT_CDS_RETURN_(false);
+  static bool is_using_stubs() NOT_CDS_RETURN_(false);
```
We have singular naming (`is_dumping_stub()`) for these methods in `premain` branch.
I was debating to do separate RFE for renaming in mainline or may be you can do it here.
It is up to you.
I did not pay attention to these when I work on adapter caching. But now I have to merge from mainline to premain and I noticed difference.

ashu-mehra replied:
I will make this change.

vnkozlov commented:
@ashu-mehra,  this looks good with few comments. After you address them, please merge latest jdk - I pushed small related change to limit platforms to run with AOT.

After that I will submit new testing.

ashu-mehra commented:
@vnkozlov addressed your comments.
I also noticed the newly added test `AOTCodeCompressedOopsTest` was consistently failing on macos-aarch64 because for some reason the CompressedOops::shift=3 even for heap size as low as for Xmx128m. So I have updated the test to make it more resilient by reading the 'actual' base and shift from Xlog:cds output.

vnkozlov reviewed (commented):
Looks good. I submitted testing.

vnkozlov reviewed (approved):
My testing passed.

ashu-mehra commented:
@adinn can you please review this as well?

ashu-mehra commented:
@vnkozlov thank you for reviewing and testing it.

adinn commented on src/hotspot/cpu/aarch64/sharedRuntime_aarch64.cpp:
```diff
@@ -2184,6 +2185,12 @@ void SharedRuntime::generate_deopt_blob() {
   }
 #endif
   const char* name = SharedRuntime::stub_name(SharedStubId::deopt_id);
+  CodeBlob* blob = AOTCodeCache::load_code_blob(AOTCodeEntry::SharedBlob, (uint)SharedStubId::deopt_id, name);
```
For most shared blobs we have a single entry at offset 0. However, for the deopt blob we have 3 (or 5) extra entry points which are embedded in the deopt blob as field values. Saving and restoring the blob internal state removes the need to pass a count and array of entry offsets into load_code_blob and store_code_blob at this point.
That makes we wonder if we need to do the same with AdapterBlob. If we embedded the offsets that are currently stored in AdapterHandlerEntry into AdapterBlob then we could also avoid having to explicitly pass the count and array of offsets at AOT load/store for adapters. The getters in AdapterHandlerEntry could fetch them indirectly or else the entry could cache them locally when it is initialized depending on whether we care about a memory indirection. Maybe this would make things more uniform?

ashu-mehra replied:
yup, I agree and I have the similar idea of storing entry points in AdapterBlob just like DeoptimizationBlob. Currently the pointer to AdapterBlob is not maintained anywhere. So the AdapterHandlerEntry would also have to maintain pointer to AdapterBlob.
I was actually wondering if we can eliminate AdapterHandlerEntry by also storing AdapterFingerprint in the AdapterBlob. The Method can then have a pointer to AdapterBlob.

vnkozlov replied:
Yes, let's do that in follow up RFE.

@adinn can you approve current changes so we can proceed?

adinn replied:
Done!

adinn reviewed (approved):
I looked through everything and could not spot any issues other than moving the adapter offsets into the blob. If we do decide to change that it can wait until we fold in the StubGen stub save/restore -- where we will need to save the blob plus extra info about the stubs within the blob and offsets to oner more entries within each stub.
We will also eventually need to clean up registration of stub entry addresses in the address table but that has to wait until I have completed implementing generator code to produce an enum that identifies all generated entries.
So I think this is ok to go in for now.

ashu-mehra commented:
@adinn thanks for the review

ashu-mehra commented:
/integrate

dholmes-ora commented:
This change has broken the Zero build - https://bugs.openjdk.org/browse/JDK-8357084
